nexus/reconfigurator/planning/src/blueprint_builder/builder/disks_editor.rs

```diff
@@ -0,0 +1,178 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Helper for editing the disks of a Blueprint
+
+use nexus_types::deployment::BlueprintPhysicalDiskConfig;
+use nexus_types::deployment::BlueprintPhysicalDisksConfig;
+use omicron_common::api::external::Generation;
+use omicron_uuid_kinds::GenericUuid;
+use omicron_uuid_kinds::PhysicalDiskUuid;
+use omicron_uuid_kinds::SledUuid;
+use std::collections::BTreeMap;
+use std::collections::BTreeSet;
+
+/// Helper for working with sets of disks on each sled
+///
+/// Tracking the set of disks is slightly non-trivial because we need to
+/// bump the per-sled generation number iff the disks are changed.  So
+/// we need to keep track of whether we've changed the disks relative
+/// to the parent blueprint.
+#[derive(Debug)]
+pub(super) struct BlueprintDisksEditor {
+    current: BTreeMap<SledUuid, DisksConfig>,
+    changed: BTreeSet<SledUuid>,
+}
+
+impl BlueprintDisksEditor {
+    pub fn new(
+        current: BTreeMap<SledUuid, BlueprintPhysicalDisksConfig>,
+    ) -> Self {
+        let current = current
+            .into_iter()
+            .map(|(sled_id, config)| (sled_id, config.into()))
+            .collect();
+        Self { current, changed: BTreeSet::new() }
+    }
+
+    /// Get a helper to edit the disks of a specific sled.
+    ///
+    /// If any changes are made via the returned editor, the sled will be
+    /// recorded as needing a generation bump in its disk config when the editor
+    /// is dropped.
+    pub fn sled_disks_editor(
+        &mut self,
+        sled_id: SledUuid,
+    ) -> SledDisksEditor<'_> {
+        let config =
+            self.current.entry(sled_id).or_insert_with(DisksConfig::empty);
+        SledDisksEditor::new(sled_id, config, &mut self.changed)
+    }
+
+    #[cfg(test)]
+    pub fn current_sled_disks(
+        &self,
+        sled_id: &SledUuid,
+    ) -> Option<&BTreeMap<PhysicalDiskUuid, BlueprintPhysicalDiskConfig>> {
+        let config = self.current.get(sled_id)?;
+        Some(&config.disks)
+    }
+
+    pub fn build(
```

Comment:
Could we include some docs about why we're passing `sled_ids` again here?

(e.g., why are we passing this here, when `BlueprintDisksEditor` is already taking a map of sleds as input?)

Reply:
I agree with Sean that a comment is a good idea.

 To check my understanding, the sleds we take in the constructor are all from the parent blueprint, but these sleds are the `InService` sleds from the `PlanningInput` which may be different. This allows us to add a `DiskConfig::Empty` for any new sleds, and also to ignore sleds that are no longer in service but may have configurations.